Added driver for cs43l22 audio DAC

vadimol opened the pull request.

vadimol committed:
Added driver for cs43l22 audio DAC

vadimol commented:
/cc @forGGe

forGGe commented on dev/cs43l22/export/dev/cs43l22.hpp:
```diff
@@ -0,0 +1,838 @@
+    {
+        single    = 0x40, // Play one beep
+        multiple  = 0x80, // Play beep for on_time, mute for off_time
+        continuos = 0xC0, // Play continuously
```
Typo: 
`continuos` -> `continuous`

vadimol replied:
thanks

forGGe commented on dev/cs43l22/export/dev/cs43l22.hpp:
```diff
@@ -0,0 +1,838 @@
+    //!  Send PCM buffer to the codec.
+    //! \details This method blocks until all data is transferred.
+    //!  The PCM data contains audio samples. Sample size is constant and equal to 16 bit.
+    //!  Sampling frequency can be changed with cs43l22< I2c, I2s, Rst_gpio >::set_sampling_frequency() method.
```
I think doxygen is fine with simple `set_sampling_frequency()` without prepended class name

vadimol replied:
This way doxygen will automatically generate a link to method

forGGe commented on dev/cs43l22/export/dev/cs43l22.hpp:
```diff
@@ -0,0 +1,838 @@
+    //! \tparam frequency Audio frequency.
+    //! \retval Status of the operation.
+    //!
+    template < ecl::i2s::audio_frequency frequency>
```
CS

vadimol replied:
Do you mean 
template  <ecl::i2s::audio_frequency frequency>
?

forGGe replied:
Yes

vadimol replied:
Done

forGGe commented on dev/cs43l22/export/dev/cs43l22.hpp:
```diff
@@ -0,0 +1,838 @@
+
+    // power up sequence according to RM
+    ecl::err rc = register_write(pwr_ctrl1, 0x01);
+    //TODO add error checking.
```
Let's add github issue for this

vadimol replied:
Ok

forGGe commented on dev/cs43l22/export/dev/cs43l22.hpp:
```diff
@@ -0,0 +1,838 @@
+    register_write(init_sequence_reg1, 0x99);
+    register_write(init_sequence_reg3, 0x80);
+
+    register_read(init_sequence_reg2, value);
```
It will be good if you add comment near each register write/read to explain clean purpose of doing these operations (besides that this is an init process in general)

vadimol replied:
Unfortunately, RM does not provide information about what does it mean. 
It goes something like this
1. Write 0x99 to register 0x00.
2. Write 0x80 to register 0x47.

forGGe replied:
Ah, I see

forGGe commented on dev/cs43l22/export/dev/cs43l22.hpp:
```diff
@@ -0,0 +1,838 @@
+}
+
+template < class I2c, class I2s, class Rst_gpio >
+ecl::err cs43l22< I2c, I2s, Rst_gpio >::send_pcm_buffer(uint16_t *buffer, size_t count)
```
Add `const` specifier, if buffer doesn't change

vadimol replied:
Yes, _const_ should be here

forGGe commented on dev/cs43l22/export/dev/cs43l22.hpp:
```diff
@@ -0,0 +1,838 @@
+
+    I2s::lock();
+
+    I2s::set_buffers((uint8_t *)buffer, nullptr, count * sizeof(*buffer));
```
Use `reinterpret_cast`, as more common way to express casts in C++

vadimol replied:
ok

forGGe commented on dev/cs43l22/export/dev/cs43l22.hpp:
```diff
@@ -0,0 +1,838 @@
+}
+
+template < class I2c, class I2s, class Rst_gpio >
+ecl::err cs43l22< I2c, I2s, Rst_gpio >::pcm_stream_start(uint16_t *buffer, size_t count, user_callback cb)
```
Same with `const`

vadimol replied:
Yes, const should be here

forGGe commented on dev/cs43l22/export/dev/cs43l22.hpp:
```diff
@@ -0,0 +1,838 @@
+    //!  To use this method, the init() method must be called previously.
+    //! \retval Status of the operation.
+    //!
+    static ecl::err power_up();
```
You can use simply `err`, since class is already in the `ecl` namespace

vadimol replied:
Done

forGGe commented on dev/cs43l22/export/dev/cs43l22.hpp:
```diff
@@ -0,0 +1,838 @@
+template < class I2c, class I2s, class Rst_gpio >
+ecl::err cs43l22< I2c, I2s, Rst_gpio >::pcm_stream_start(uint16_t *buffer, size_t count, user_callback cb)
+{
+    ecl_assert(m_state & inited);
```
Add assertion here that `user_callback` contains valid callable object:
http://en.cppreference.com/w/cpp/utility/functional/function/operator_bool

vadimol replied:
Done

forGGe commented:
Partial review

vadimol commented on dev/cs43l22/export/dev/cs43l22.hpp:
```diff
@@ -0,0 +1,838 @@
+    //! \param ch Audio channel.
+    //! \retval Status of the operation.
+    //!
+    static ecl::err headphone_mute_on(channel ch = channel::all);
```
yes, it will be better, thanks

vadimol commented on dev/cs43l22/export/dev/cs43l22.hpp:
```diff
@@ -0,0 +1,838 @@
+    //! \param mode The operation mode for beep generator.
+    //! \retval Status of the operation.
+    //!
+    static ecl::err beep_generator_enable(beep_mode mode = beep_mode::multiple);
```
yes, it will be better, thanks

forGGe commented on dev/cs43l22/export/dev/cs43l22.hpp:
```diff
@@ -0,0 +1,838 @@
+template < class I2c, class I2s, class Rst_gpio >
+ecl::err cs43l22< I2c, I2s, Rst_gpio >::pcm_stream_stop()
+{
+    ecl_assert(m_state & inited);
```
Here. we need to ensure that `pcm_stream_start()` is actually called previously and stream wasn't stopped yet.

vadimol replied:
Done

forGGe commented on dev/cs43l22/export/dev/cs43l22.hpp:
```diff
@@ -0,0 +1,838 @@
+    ecl_assert(m_state & inited);
+
+    if (!buffer || !count) {
+        return ecl::err::inval;
```
Let's discuss this.

vadimol replied:
Discussed and updated

vadimol committed:
The CS43l22 driver:
1) Updated description
2) Improved documentation
3) Added asserts
4) Fixed code style issues

vadimol commented:
/cc @forGGe 

addressed review comments.
